engines/freescape/freescape.cpp

```diff
@@ -0,0 +1,609 @@
+Math::Vector3d FreescapeEngine::directionToVector(float pitch, float heading) {
+	Math::Vector3d v;
+
+	float radHeading = Common::deg2rad(heading);
+	float radPitch = Common::deg2rad(pitch);
+
+	v.setValue(0, cos(radPitch) * cos(radHeading));
+	v.setValue(1, sin(radPitch));
+	v.setValue(2, cos(radPitch) * sin(radHeading));
+	v.normalize();
+
+	return v;
+}
+
+void FreescapeEngine::drawUI() {
+	_gfx->renderCrossair(0, _crossairPosition);
+	_gfx->setViewport(_viewArea);
+}
+
+void FreescapeEngine::drawFrame() {
+	_gfx->updateProjectionMatrix(60.0, _nearClipPlane, _farClipPlane);
+	_gfx->positionCamera(_position, _position + _cameraFront);
+	_currentArea->draw(_gfx);
+	drawBorder();
+	drawUI();
+}
+
+void FreescapeEngine::pressedKey(const int keycode) {}
+
+void FreescapeEngine::generateInput() {
+	Common::Event event;
+	if (isDOS()) {
+
+		if (_currentDemoInputRepetition == 0) {
+			_currentDemoInputRepetition = 1;
+			_currentDemoInputCode = _demoData[_demoIndex++];
+			if (_currentDemoInputCode >= 0x80) {
+				_currentDemoInputRepetition = _currentDemoInputCode - 0x80;
+				assert(_currentDemoInputRepetition > 0);
+				_currentDemoInputCode = _demoData[_demoIndex++];
+			}
+		}
+
+		if (_currentDemoInputCode >= 0x16 && _currentDemoInputCode <= 0x1a) {
+			// 0x16 -> left click / shoot
+			// 0x17 -> right?
+			// 0x18 -> left
+			// 0x19 -> down?
+			// 0x1a -> up
+			// TODO: mouse events
+		} else if (_currentDemoInputCode == 0x7f) {
+			// TODO: wait?
+		} else {
+			event = Common::Event();
+			event.type = Common::EVENT_KEYDOWN;
+			event.kbd.keycode = (Common::KeyCode)decodeDOSKey(_currentDemoInputCode);
+			event.customType = 0xde00;
+			g_system->getEventManager()->pushEvent(event);
+			debugC(1, kFreescapeDebugMove, "Pushing key: %x with repetition %d", event.kbd.keycode, _currentDemoInputRepetition);
+		}
+		_currentDemoInputRepetition--;
+		g_system->delayMillis(50);
+		return;
+	}
+
+	int mouseX = _demoData[_demoIndex++] << 1;
+	int mouseY = _demoData[_demoIndex++];
+	debugC(1, kFreescapeDebugMove, "Mouse moved to: %d, %d", mouseX, mouseY);
+
+	event.type = Common::EVENT_MOUSEMOVE;
+	event.mouse = Common::Point(mouseX, mouseY);
+	event.customType = 0xde00;
+	g_system->getEventManager()->pushEvent(event);
+
+	byte nextKeyCode = _demoData[_demoIndex++];
+
+	if (nextKeyCode == 0x30) {
+		event.type = Common::EVENT_LBUTTONDOWN; // Keep same event fields
+		g_system->getEventManager()->pushEvent(event);
+		nextKeyCode = _demoData[_demoIndex++];
+	} else {
+		while (nextKeyCode != 0) {
+			event = Common::Event();
+			event.type = Common::EVENT_KEYDOWN;
+			event.kbd.keycode = (Common::KeyCode)decodeAmigaAtariKey(nextKeyCode);
+			debugC(1, kFreescapeDebugMove, "Pushing key: %x", event.kbd.keycode);
+			event.customType = 0xde00;
+			g_system->getEventManager()->pushEvent(event);
+			nextKeyCode = _demoData[_demoIndex++];
+		}
+	}
+	assert(!nextKeyCode);
+	g_system->delayMillis(50);
```

Comment:
You're artificially restrict it to 20FPS. Is this intended?

Reply:
Yes, the DOS release of the game original runs in 10FPS in fact. I believe this can be increased to without affecting the rest of the game (it was the most popular complain with the original game). This code is part of the demo replay, which is a sequence of events to in keyboard and mouse. Since this part is non-interactive, this will be capped at 10FPS to match the original implementation.